### PR TITLE
Cudnn base

### PR DESCRIPTION
**Pull request by M1ngXU**
https://github.com/coreylowman/cudarc/issues/18#issuecomment-1375928870

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -22,4 +23,5 @@ pub mod prelude {
         AsKernelParam, CudaDevice, CudaDeviceBuilder, CudaSlice, DriverError, LaunchAsync,
         LaunchConfig, ValidAsZeroBits,
     };
+    pub use crate::cudnn::*;
```
Comment (coreylowman) [inline src/lib.rs:26]: Let's not include this in the prelude for now

File: src/cudnn/bindgen.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  --whitelist-function="^cudnn.*" \
+  --default-enum-style=rust \
+  --no-doc-comments \
+  --with-derive-default \
+  --with-derive-eq \
+  --with-derive-hash \
+  --with-derive-ord \
+  --size_t-is-usize \
+  --use-core \
+  wrapper.h -- --include-directory "/mnt/e/Max/cudnn/include" --include-directory "/mnt/c/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include" \
```
Comment (coreylowman) [inline src/cudnn/bindgen.sh:16]:
```suggestion
  wrapper.h -- -I/usr/local/cuda/include \
```
to keep standard with other bindgens. I guess this doesn't work on windows though?

File: build.rs [hunk 3]
```diff
@@ -5,4 +5,6 @@ fn main() {
     println!("cargo:rustc-link-lib=dylib=cudart");
     println!("cargo:rustc-link-lib=dylib=cublas");
     println!("cargo:rustc-link-lib=dylib=cublasLt");
+    println!("cargo:rustc-link-search=native={}", std::env::var("CUDNN_LIB").expect("Failed to read environmental variable `CUDNN_LIB`. Set `CUDNN_LIB` to `path/to/cudnn/lib`."));
```
Comment (coreylowman) [inline build.rs:8]: Do you know why this is needed for cudnn but not other libs?

File: .gitignore [hunk 4]
```diff
@@ -1,2 +1,3 @@
 /target
 /Cargo.lock
+cudnn.log
```
Comment (coreylowman) [inline .gitignore:3]: Is this something cudnn creates?

**Review by coreylowman — COMMENTED** [review]
Looks great! Will approve after cleaning up some of the paths that are specific to your machine it looks like

File: src/lib.rs (hunk 1 above)
Comment (M1ngXU) [reply src/lib.rs:26]: ok

File: src/cudnn/bindgen.sh (hunk 2 above)
Comment (M1ngXU) [reply src/cudnn/bindgen.sh:16]: hmm, this is in wsl since bindgen doesn't work on windows for me

File: build.rs (hunk 3 above)
Comment (M1ngXU) [reply build.rs:8]: because cudnn is installed separately and is not in the folder where cuda resides

File: .gitignore (hunk 4 above)
Comment (M1ngXU) [reply .gitignore:3]: i have an env var to log all cudnn calls, i can remove that if necessary

File: build.rs (hunk 3 above)
Comment (coreylowman) [reply build.rs:8]: Ooooh did not realize that

File: src/cudnn/result.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for CudnnError {}
+
+// No sync because of https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#thread-safety
+pub struct CudnnHandle(pub(crate) cudnnHandle_t);
```
Comment (coreylowman) [inline src/cudnn/result.rs:35]: This needs to keep a reference to Arc<CudaDevice> to make sure this doesn't outlive device:
```suggestion
pub struct CudnnHandle {
    pub(crate) handle: cudnnHandle_t,
    pub(crate) device: Arc<CudaDevice>,
}
```

And then `fn create` should take an `Arc<CudaDevice>`.

Did a similar thing with blas if you want an example

File: build.rs (hunk 3 above)
Comment (coreylowman) [reply build.rs:8]: So we'll probably need to wrap this functionality with a feature if it's not installed by default with cuda. Will investigate a bit on my dev machine to see if it's installed/what the process is like. For dfdx I wonder if we should require cudnn, or fallback to custom kernels if people don't have cudnn installed

Comment (M1ngXU) [reply build.rs:8]: does something like pytorch or tensorflow work with cuda but without cudnn? i don't think so and i don't see any reason why someone could install cuda but not cudnn ...

File: src/cudnn/result.rs (hunk 5 above)
Comment (M1ngXU) [reply src/cudnn/result.rs:35]: ok, didn't think of that 😅 

by the way, the test "launch with 16 bit" fails for me with an buffer overflow; are you having the same issue?

Comment (coreylowman) [reply src/cudnn/result.rs]: I haven't, I can check again

File: build.rs (hunk 3 above)
Comment (coreylowman) [reply build.rs:8]: I think pytorch bundles cuda/cudnn libraries together? Not completely sure tbh. If cudnn is installed by default then i don't think we need feature flag. We can just require it but I want to go through install process to see how annoying it is first

File: src/cudnn/result.rs (hunk 5 above)
Comment (M1ngXU) [reply src/cudnn/result.rs]: ok thanks

File: build.rs (hunk 3 above)
Comment (M1ngXU) [reply build.rs:8]: not sure about pytorch, but it is seperate in tensorflow (you have to install everything yourself anyway)...

**Comment by coreylowman** [comment]
Closing since this was merged in #97

**Pull request closed without merging**